vpr/src/place/net_cost_handler.cpp

```diff
@@ -0,0 +1,2277 @@
+ * them as chan?_place_cost_fac[subhigh][sublow].  They are used to     *
+ * speed up the computation of the cost function that takes the length  *
+ * of the net bounding box in each dimension, divided by the average    *
+ * number of tracks in that direction; for other cost functions they    *
+ * will never be used.                                                  *
+ */
+static vtr::NdMatrix<float, 2> chanx_place_cost_fac({0, 0}); //[0...device_ctx.grid.width()-2]
+static vtr::NdMatrix<float, 2> chany_place_cost_fac({0, 0}); //[0...device_ctx.grid.height()-2]
+
+/* Cost of a net, and a temporary cost of a net used during move assessment. */
+static vtr::vector<ClusterNetId, double> net_cost, proposed_net_cost;
+
+/* [0...cluster_ctx.clb_nlist.nets().size()-1]                                               *
+ * A flag array to indicate whether the specific bounding box has been updated   *
+ * in this particular swap or not. If it has been updated before, the code       *
+ * must use the updated data, instead of the out-of-date data passed into the    *
+ * subroutine, particularly used in try_swap(). The value NOT_UPDATED_YET        *
+ * indicates that the net has not been updated before, UPDATED_ONCE indicated    *
+ * that the net has been updated once, if it is going to be updated again, the   *
+ * values from the previous update must be used. GOT_FROM_SCRATCH is only        *
+ * applicable for nets larger than SMALL_NETS and it indicates that the          *
+ * particular bounding box cannot be updated incrementally before, hence the     *
+ * bounding box is got from scratch, so the bounding box would definitely be     *
+ * right, DO NOT update again.                                                   */
+static vtr::vector<ClusterNetId, NetUpdateState> bb_updated_before;
+
+/* The following arrays are used by the try_swap function for speed.   */
+
+/**
+ * The wire length estimation is based on the bounding box of the net. In the case of the 2D architecture,
+ * we use a 3D BB with the z-dimension (layer) set to 1. In the case of 3D architecture, there 2 types of bounding box:
+ * 3D and per-layer. The type is determined at the beginning of the placement and stored in the placement context.
+ *
+ *
+ * If the bonding box is of the type 3D, ts_bb_coord_new and ts_bb_edge_new are used. Otherwise, layer_ts_bb_edge_new and
+ * layer_ts_bb_coord_new are used.
+ */
+
+/* [0...cluster_ctx.clb_nlist.nets().size()-1] -> 3D bounding box*/
+static vtr::vector<ClusterNetId, t_bb> ts_bb_coord_new, ts_bb_edge_new;
+/* [0...cluster_ctx.clb_nlist.nets().size()-1][0...num_layers] -> 2D bonding box on a layer*/
+static vtr::vector<ClusterNetId, std::vector<t_2D_bb>> layer_ts_bb_edge_new, layer_ts_bb_coord_new;
+/* [0...cluster_ctx.clb_nlist.nets().size()-1][0...num_layers] -> number of sink pins on a layer*/
+static vtr::Matrix<int> ts_layer_sink_pin_count;
+/* [0...num_afftected_nets] -> net_id of the affected nets */
```

Comment:
The size of this data structure is initialized to the number of nets at the beginning of the placement to reduce the runtime required for memory allocation and deallocation. Because of this, `num_affected_nets` is passed to the functions alongside this data structure, and the elements up to that number are valid.

Reply:
OK.  You can leave it then but comment it well.  I think it would be slightly more clear to put num_affected_nets in a struct with the ts_* variables so they are all grouped then; passing one variable and using file scope ones for other closely related data seems odd.